### DISCUSSION TITLE
Param's function not narrowed base on the `@overload` signatures

### DISCUSSION THREAD

**Discussion by karimelhajoui63** [microsoft/pyright#7200] [General]
Hi, 

I'm not sure if it's a missing feature or if it's intentional, but the param's function aren't narrowed based on the `@overload` signatures.
It seems that the `@overload` signatures are used only for the caller side while it seems possible to narrow down on the function's definition.

For exemple, in the exemple, I was expecting `int` within the `if` and `str` within the `elif`:

```py
from typing import Literal, overload

@overload
def func(type_: Literal["type_int"], param: int) -> None: ...

@overload
def func(type_: Literal["type_str"], param: str) -> None: ...

def func(type_: Literal["type_int", "type_str"], param: int | str) -> None:
    if type_ == "type_int":
        reveal_type(param)  # Type of "param" is "int | str"
    elif type_ == "type_str":
        reveal_type(param)  # Type of "param" is "int | str"
```

I'm using `Pylance v2023.12.1`

**Comment by erictraut** [comment]
This isn't a missing feature. This is how `@overload` works according to the typing spec.

The `@overload` signatures apply only to callers of the function. The implementation must provide its own polymorphic logic and parameter validation. You would typically include checks like this:

```python
def func(type_: Literal["type_int", "type_str"], param: int | str) -> None:
    if type_ == "type_int":
        if not isinstance(param, int):
            raise ValueError("Expected int!")
        reveal_type(param) # Type of "param" is str
    elif type_ == "type_str":
        if not isinstance(param, str):
            raise ValueError("Expected str!")
        reveal_type(param)  # Type of "param" is "int"
```

**Reply by karimelhajoui63** [reply]
Ok, thanks.

So, this solution implies code duplication if I want a type safe environment on the called **and** definition side 😕 but it seems that there is not other solution currently available...
